src/librustc_mir/borrow_check/nll/type_check/relate_tys.rs

```diff
@@ -0,0 +1,510 @@
+                self.push_outlives(v_b, v_a);
+            }
+
+            if self.ambient_contravariance() {
+                // Contravariant: b <= a. Hence, `a: b`.
+                self.push_outlives(v_a, v_b);
+            }
+        }
+
+        Ok(a)
+    }
+
+    fn binders<T>(
+        &mut self,
+        a: &ty::Binder<T>,
+        b: &ty::Binder<T>,
+    ) -> RelateResult<'tcx, ty::Binder<T>>
+    where
+        T: Relate<'tcx>,
+    {
+        // We want that
+        //
+        // ```
+        // for<'a> fn(&'a u32) -> &'a u32 <:
+        //   fn(&'b u32) -> &'b u32
+        // ```
+        //
+        // but not
+        //
+        // ```
+        // fn(&'a u32) -> &'a u32 <:
+        //   for<'b> fn(&'b u32) -> &'b u32
+        // ```
+        //
+        // We therefore proceed as follows:
+        //
+        // - Instantiate binders on `b` universally, yielding a universe U1.
+        // - Instantiate binders on `a` existentially in U1.
+
+        debug!(
+            "binders({:?}: {:?}, ambient_variance={:?})",
+            a, b, self.ambient_variance
+        );
+
+        if self.ambient_covariance() {
+            // Covariance, so we want `for<'a> A <: for<'b...> B` --
+            // therefore we replace all of the `'b...` regions with
+            // placeholders and then replace `'a...` with
+            // existentials. We then check if any instantiation of A
+            // can match against the placeholders in B.
+
+            let b_scope = self.create_scope(b, UniversallyQuantified(true));
+            let a_scope = self.create_scope(a, UniversallyQuantified(false));
+
+            debug!("binders: a_scope = {:?} (existential)", a_scope);
+            debug!("binders: b_scope = {:?} (universal)", b_scope);
+
+            self.b_scopes.push(b_scope);
+            self.a_scopes.push(a_scope);
+
+            // FIXME -- to be fully correct, we would set the ambient
+            // variance to Covariant here. As is, we will sometimes
+            // propagate down an ambient variance of Equal -- this in
+            // turn causes us to report errors in some cases where
+            // types perhaps *ought* to be equal. See the
+            // `hr-fn-aau-eq-abu.rs` test for an example. Fixing this
+            // though is a bit nontrivial: in particular, it would
+            // require a more involved handling of canonical
+            // variables, since we would no longer be able to rely on
+            // having an `==` relationship for canonical variables.
+
+            self.relate(a.skip_binder(), b.skip_binder())?;
+
+            self.b_scopes.pop().unwrap();
+            self.a_scopes.pop().unwrap();
+        }
+
+        if self.ambient_contravariance() {
+            // Contravariance, so we want `for<'a> A :> for<'b...> B`
+            // -- do the opposite, therefore, of covariant. Replace
+            // the `'a...` placeholders with `'b...` variables.
```

Comment:
I find the phrasing "replace the `'a...` placeholders with `'b...` variables" confusing.

Would it be acceptable to instead say "check if any instantiation of B (where `'b...` have been replaced with existentials) can match against any instantiation of A (where `'a...` have been replaced with placeholders)", which more closely matches the wording you used above (which I understand better)?

Or is there some reason you chose this particular phrasing here?

Reply:
Your way seems clearer, will fix.